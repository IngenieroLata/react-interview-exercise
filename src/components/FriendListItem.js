import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './FriendListItem.css';

class FriendListItem extends Component {
  render() {
    const { id, name, gender, starFriend, starred, deleteFriend } = this.props;

    return (
      <li className={styles.friendListItem}>
        <div className={styles.friendInfos}>
          <div data-testid="name">
            <span>{name}</span>
          </div>
          <div>
            <small>xx friends in common</small>
          </div>
          {gender && (
            <div data-testid="gender">
              <small>{gender}</small>
            </div>
          )}
        </div>
        <div className={styles.friendActions}>
          <button
            data-testid="star"
            className={`btn btn-default ${styles.btnAction}`}
            onClick={() => starFriend(id)}>
            <i
              className={classnames('fa', {
                'fa-star': starred,
                'fa-star-o': !starred
              })}
            />
          </button>
          <button
            data-testid="delete"
            className={`btn btn-default ${styles.btnAction}`}
            onClick={() => deleteFriend(id)}>
            <i className="fa fa-trash" />
          </button>
        </div>
      </li>
    );
  }
}

FriendListItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  starred: PropTypes.bool,
  gender: PropTypes.string,
  starFriend: PropTypes.func.isRequired,
  deleteFriend: PropTypes.func.isRequired
};

FriendListItem.defaultProps = {
  starred: false,
  gender: ''
};

export default FriendListItem;
