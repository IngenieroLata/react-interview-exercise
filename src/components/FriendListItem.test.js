import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import FriendListItem from './FriendListItem';

const props = {
  id: 1,
  name: 'Jane Doe',
  starred: true,
  gender: 'female',
  starFriend: jest.fn(),
  deleteFriend: jest.fn()
};

describe('<FriendListItem />', () => {
  it('Render the correct Name', () => {
    const wrapper = shallow(<FriendListItem {...props} />);
    expect(wrapper.find('[data-testid="name"]').text()).toBe('Jane Doe');
  });

  it('Render the correct Gender', () => {
    const wrapper = shallow(<FriendListItem {...props} />);
    expect(wrapper.find('[data-testid="gender"]').text()).toBe('female');
  });

  it('Render the mark as active star', () => {
    const wrapper = shallow(<FriendListItem {...props} />);
    expect(wrapper.find('.fa-star')).toHaveLength(1);
  });

  it('Call starFriend action', () => {
    const wrapper = shallow(<FriendListItem {...props} />);
    wrapper.find('[data-testid="star"]').simulate('click');
    expect(props.starFriend).toHaveBeenCalledWith(props.id);
  });

  it('Call deleteFriend action', () => {
    const wrapper = shallow(<FriendListItem {...props} />);
    wrapper.find('[data-testid="delete"]').simulate('click');
    expect(props.deleteFriend).toHaveBeenCalledWith(props.id);
  });

  it('Matches the snapshot render', () => {
    const wrapper = shallow(<FriendListItem {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
