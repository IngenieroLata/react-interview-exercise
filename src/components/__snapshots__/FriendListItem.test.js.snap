exports[`<FriendListItem /> Matches the snapshot render 1`] = `
<li>
  <div>
    <div
      data-testid="name">
      <span>
        Jane Doe
      </span>
    </div>
    <div>
      <small>
        xx friends in common
      </small>
    </div>
    <div
      data-testid="gender">
      <small>
        female
      </small>
    </div>
  </div>
  <div>
    <button
      className="btn btn-default undefined"
      data-testid="star"
      onClick={[Function]}>
      <i
        className="fa fa-star" />
    </button>
    <button
      className="btn btn-default undefined"
      data-testid="delete"
      onClick={[Function]}>
      <i
        className="fa fa-trash" />
    </button>
  </div>
</li>
`;
